### PR TITLE
Fax Machine

### PR DESCRIPTION
**Pull request by cerebral-314**
## About The Pull Request
Adds a fax machine for the bridge and Lawyer's Office.

Ported from Para/Bay
## Why It's Good For The Game
Imagine that you're the lawyer, and you notice HoS jails someone and follows out an execution that the Captain didn't give permission for. This will give the lawyer the ability to get the HoS punished/fired. 

Essentially, this gives the Lawyer an opportunity to actually do their job.

## Changelog
:cl:
add: fax machine
soundadd: fax sound
imageadd: fax machine icons
spellcheck: fixed a few typos
admin: admin tools for faxes
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
kind of just an IC ahelp machine.

**Comment by cerebral-314** [comment]
> 
> 
> kind of just an IC ahelp machine.

It's mostly for Sage... it's kinda weird praying for an admeme to fire the HoS.
Also it was requested a shitload...

**Comment by OceanFish1** [comment]
*nervously typing in fax* I saw captain erping in dorms pls demote he i was muted from aehlp

File: code/modules/admin/topic.dm [hunk 1]
```diff
@@ -1223,6 +1223,230 @@
 
 		usr.client.smite(H)
 
+	else if(href_list["FaxReplyTemplate"])
+		if(!check_rights(R_ADMIN))
```
Comment (Archanial) [inline code/modules/admin/topic.dm]: Since smiting is fun, I think this should be fun too.

File: code/modules/admin/topic.dm [hunk 2]
```diff
@@ -1223,6 +1223,230 @@
...
+			return
+		var/mob/living/carbon/human/H = locate(href_list["FaxReplyTemplate"]) in GLOB.mob_list
```
Comment (Archanial) [inline code/modules/admin/topic.dm]: I'm pretty sure there is global list of players

File: .gitignore [hunk 3]
```diff
@@ -216,3 +216,6 @@ tools/MapAtmosFixer/MapAtmosFixer/bin/*
 !/config/jukebox_music/sounds/exclude
 /config/title_music/sounds/*
 !/config/title_music/sounds/exclude
+config/admins.txt
+config/config.txt
+config/dbconfig.txt
```
Comment (Archanial) [inline .gitignore]: It's not intentional, is it?

File: code/__DEFINES/say.dm [hunk 4]
```diff
@@ -89,6 +89,9 @@
 #define MAX_BROADCAST_LEN		512
 #define MAX_CHARTER_LEN			80
 
+#define MAX_PAPER_MESSAGE_LEN 3072
+#define MAX_PAPER_FIELDS 50
```
Comment (Archanial) [inline code/__DEFINES/say.dm]: Use the defines located in paper.dm or move them all here

File: code/modules/admin/topic.dm [hunk 5]
```diff
@@ -1223,6 +1223,230 @@
...
+		if(!H || !istype(H))
+			to_chat(usr, "<span class='warning'>This can only be used on instances of type /mob/living/carbon/human</span>")
+			return
+		var/obj/item/paper/P = new /obj/item/paper(null)
```
Comment (Archanial) [inline code/modules/admin/topic.dm]:
```suggestion
		var/obj/item/paper/P = new /obj/item/paper()
```

File: code/modules/admin/topic.dm [hunk 6]
```diff
@@ -1223,6 +1223,230 @@
...
+		else if(stype == "You are wasting our time")
+			tmsg += "Greetings, esteemed crewmember. Your fax has been <b><I>DECLINED</I></b> automatically by NAS Trurl Fax Registration.<BR><BR>In the interest of preventing further mismanagement of company resources, please avoid wasting our time with such petty drivel.<BR><BR>Do kindly remember that we expect our workforce to maintain at least a semi-decent level of profesionalism. Do not test our patience.<BR><BR><i><small>This is an automatic message.</i></small>"
+		else if(stype == "Keep up the good work")
+			tmsg += "Greetings, esteemed crewmember. Your fax has been received successfully by NAS Trurl Fax Registration.<BR><BR>We at the NAS Trurl appreciate the good work that you have done here, and sincerely recommend that you continue such a display of dedication to the company.<BR><BR><i><small>This is absolutely not an automated message.</i></small>"
```
Comment (Archanial) [inline code/modules/admin/topic.dm]: use switch instead

File: code/modules/admin/topic.dm [hunk 7]
```diff
@@ -1223,6 +1223,230 @@
...
+		P.info = tmsg
+		P.x = rand(-2, 0)
+		P.y = rand(-1, 2)
+		P.offset_x += P.x
+		P.offset_y += P.y
+		P.update_icon()
+		if(!P.stamped)
+			P.stamped = new
+		if (isnull(P.stamped))
+			P.stamped = list()
```
Comment (Archanial) [inline code/modules/admin/topic.dm]: I'm not entirely sure what you mean by this code. You initilize stamped but then check if it's still null?

File: code/modules/admin/topic.dm [hunk 8]
```diff
@@ -1223,6 +1223,230 @@
...
+		stampoverlay.pixel_x = rand(-2, 2)
+		stampoverlay.pixel_y = rand(-3, 2)
+		stampoverlay.icon_state = "paper_stamp-[stampvalue]"
+		P.add_overlay(stampoverlay)
+		LAZYADD(P.stamped, "stamp-[stampvalue]")
+		if (isnull(P.stamps))
+			P.stamps = list()
+		P.stamps[++P.stamps.len] = list("paper121x54 stamp-[stampvalue]", 280.5, 3.5, 0)
+		fax.receivefax(P)
+		if(istype(H) && H.stat == CONSCIOUS && (istype(H.ears, /obj/item/radio/headset)))
```
Comment (Archanial) [inline code/modules/admin/topic.dm]: You also need to check if owner is not deaf and if headset is functioning

File: code/modules/admin/topic.dm [hunk 9]
```diff
@@ -1223,6 +1223,230 @@
...
+		if(!check_rights(R_ADMIN))
+			return
+
+		var/mob/sender = locate(href_list["AdminFaxCreate"])
+		var/obj/machinery/photocopier/faxmachine/fax = locate(href_list["originfax"])
+		var/faxtype = href_list["faxtype"]
+		var/reply_to = locate(href_list["replyto"])
+		var/destination
+		var/notify
+		var/obj/item/paper/P = new /obj/item/paper(null)
```
Comment (Archanial) [inline code/modules/admin/topic.dm]:
```suggestion
		var/obj/item/paper/P = new /obj/item/paper()
```

File: code/modules/paperwork/fax.dm [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+GLOBAL_LIST_EMPTY(adminfaxes)
+
+/datum/fax
+	var/name = "fax"
+	var/from_department = null
+	var/to_department = null
+	var/origin = null
+	var/message = null
+	var/sent_by = null
+	var/sent_at = null
```
Comment (Archanial) [inline code/modules/paperwork/fax.dm]:
```suggestion
	var/from_department
	var/to_department
	var/origin
	var/message
	var/sent_by
	var/sent_at
```

File: code/modules/paperwork/faxmachine.dm [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+	/// Can we send messages off-station?
+	var/long_range_enabled = FALSE
+	req_one_access = list(ACCESS_LAWYER, ACCESS_HEADS, ACCESS_ARMORY)
+
+	use_power = IDLE_POWER_USE
+	idle_power_usage = 30
+	active_power_usage = 200
+
+	/// ID card inserted into the machine, used to log in with
+	var/obj/item/card/id/scan = null
```
Comment (Archanial) [inline code/modules/paperwork/faxmachine.dm]:
```suggestion
	var/obj/item/card/id/scan
```

File: code/modules/paperwork/faxmachine.dm [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+
+	var/emagged = FALSE
```
Comment (Archanial) [inline code/modules/paperwork/faxmachine.dm]: Use emaged flag instead

File: code/modules/paperwork/faxmachine.dm [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+		req_one_access = list()
+		to_chat(user, "<span class='notice'>The transmitters realign to an unknown source!</span>")
+	else
+		to_chat(user, "<span class='warning'>You swipe the card through [src], but nothing happens.</span>")
+
+/obj/machinery/photocopier/faxmachine/proc/is_authenticated(mob/user)
+	if(authenticated)
+		return TRUE
+	else if(user.has_unlimited_silicon_privilege)
+		return TRUE
```
Comment (Archanial) [inline code/modules/paperwork/faxmachine.dm]:
```suggestion
	if(authenticated || user.has_unlimited_silicon_privilege)
		return TRUE
```

**Comment by ike709** [comment]
I haven't looked at this but if this is a straight port with no code improvements then I'll tell you now that it won't be merged.

File: code/__HELPERS/text.dm [hunk 14]
```diff
@@ -840,3 +845,140 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
 			continue
 		out += prob(replaceprob) ? pick(replacementchars) : char
 	return out.Join("")
+
+
+// Pencode
+/proc/pencode_to_html(text, mob/user, obj/item/pen/P = null, format = 1, sign = 1, fields = 1, deffont = PEN_FONT, signfont = SIGNFONT, crayonfont = CRAYON_FONT, no_font = FALSE)
```
Comment (francinum) [inline code/__HELPERS/text.dm]: We use markdown for paper handling.

File: .gitignore (hunk 3 above)
Comment (francinum) [inline .gitignore]: This is bad, Don't do this.

**Review by francinum — CHANGES_REQUESTED** [review]
This code is overall really, really just, *weird*

The fax reply system looks incredibly painful to use, being just a series of `alert()s`, This probably justifies it's own specialized panel.

**Comment by eLeCtrOssSnake** [comment]
Cool addition, though please do not limit it only for sage.

**Review by ike709 — CHANGES_REQUESTED** [review]
This code is just... Not good. It needs a pretty hefty rewrite.

**Comment by cerebral-314** [comment]
Will rewrite and open another PR

**Comment by AnCopper** [comment]
> Will rewrite and open another PR

Famous last words.

**Pull request closed without merging**